File: python-algo/leetcode/problem_275.py
"""275. H-Index II
https://leetcode.com/problems/h-index-ii/
"""
from typing import List
import bisect


class Solution:
    def h_index(self, citations: List[int]) -> int:
        lo, hi = 0, citations[-1]
        while lo < hi:
            mid = (lo + hi + 1) >> 1  # ceil
            pos = bisect.bisect_left(citations, mid)
            if len(citations) - pos >= mid:
                if hi == mid:
                    return mid
                lo = mid
            else:
                hi = mid - 1
        return hi

File: java-algo/src/main/java/com/leetcode/Problem275.java
package com.leetcode;

/**
 * 275. H-Index II
 * https://leetcode.com/problems/h-index-ii
 */
public class Problem275 {
    public int hIndex(int[] citations) {
        int ans = 0;
        int n = citations.length;
        int lo = citations[0], hi = citations[n - 1], mid;
        while (lo <= hi) {
            mid = (lo + hi) >> 1;
            int cnt = n - binarySearch(citations, mid);
            if (cnt >= mid) {
                ans = Math.max(ans, mid);
                lo = mid + 1;
            } else {
                ans = Math.max(ans, cnt);
                hi = mid - 1;
            }
        }
        return ans;
    }

    private int binarySearch(int[] nums, int target) {
        int lo = 0, hi = nums.length - 1, mid;
        while (lo < hi) {
            mid = (lo + hi) / 2;
            if (nums[mid] >= target) {
                hi = mid;
            } else {
                lo = mid + 1;
            }
        }
        return lo;
    }
}
